src/BlockchainWriter/ClaimController.ts

```diff
@@ -48,90 +45,89 @@ export class ClaimController {
     })
   }
 
-  async timestampNextHash() {
-    const logger = this.logger.child({ method: 'timestampNextHash' })
+  async anchorNextIPFSDirectoryHash() {
+    const logger = this.logger.child({ method: 'anchorNextHash' })
 
-    logger.trace('Retrieving Next Hash To Timestamp')
+    logger.trace('Retrieving Next Hash To Anchor')
 
     const entry = await this.collection.findOne({ txId: null })
     const ipfsDirectoryHash = entry && entry.ipfsDirectoryHash
 
-    this.logger.trace({ ipfsDirectoryHash }, 'Next Hash To Timestamp Retrieved')
+    this.logger.trace({ ipfsDirectoryHash }, 'Next IPFS Directory Hash To Anchor Retrieved')
 
     if (!ipfsDirectoryHash) return
 
     try {
-      await this.timestamp(ipfsDirectoryHash)
+      await this.anchorIPFSDirectoryHash(ipfsDirectoryHash)
     } catch (exception) {
       logger.warn(
         {
           ipfsDirectoryHash,
           exception,
         },
-        'Uncaught Exception While Timestamping Hash'
+        'Unexpected Exception While Anchoring IPFS Directory Hash'
       )
     }
   }
 
-  private async timestamp(ipfsDirectoryHash: string): Promise<void> {
-    const logger = this.logger.child({ method: 'timestamp' })
+  private async anchorIPFSDirectoryHash(ipfsDirectoryHash: string): Promise<void> {
+    const { configuration, collection, messaging, anchorData } = this
+    const logger = this.logger.child({ method: 'anchorIPFSDirectoryHash' })
+
+    logger.debug({ ipfsDirectoryHash }, 'Anchoring IPFS Hash')
 
-    logger.trace({ ipfsDirectoryHash }, 'Timestamping IPFS Hash')
+    const ipfsDirectoryHashToBitcoinData = getData(configuration.poetNetwork, configuration.poetVersion)
 
-    const utxo = await this.insightHelper.getUtxo(this.configuration.bitcoinAddress)
+    const data = ipfsDirectoryHashToBitcoinData(ipfsDirectoryHash)
+    const txId = await anchorData(data)
+
+    await collection.updateOne({ ipfsDirectoryHash }, { $set: { txId } }, { upsert: true })
+    await messaging.publish(Exchange.IPFSHashTxId, {
+      ipfsDirectoryHash,
+      txId,
+    })
+  }
 
-    if (!utxo || !utxo.length)
-      throw new Error(`Wallet seems to be empty. Check funds for ${this.configuration.bitcoinAddress}`)
+  private anchorData = async (data: string) => {
```

Comment:
I experimented with `pipeP` for this but decided not to use it yet.

I've left the experiments in [another branch](https://github.com/poetapp/node/compare/bitcoin-core-writer...blockchain-writer-functional-experiments?expand=1).

Reply:
Agreed, it probably requires to much refactoring to use it nicely atm.